openstack/compute/v2/hypervisors/results.go

```diff
@@ -0,0 +1,170 @@
+	// The memory used in this hypervisor in MB
+	MemoryMBUsed int `json:"memory_mb_used"`
+	// The number of running vms on this hypervisor
+	RunningVMs int `json:"running_vms"`
+	// The hypervisor service object
+	Service `json:"service"`
+	// The number of vcpu in this hypervisor
+	VCPUs int `json:"vcpus"`
+	// The number of vcpu used in this hypervisor
+	VCPUsUsed int `json:"vcpus_used"`
+}
+
+func (h *Hypervisor) UnmarshalJSON(b []byte) error {
+
+	type tmp Hypervisor
+	var hypervisor *struct {
+		tmp
+		CpuInfo           interface{} `json:"cpu_info"`
+		HypervisorVersion interface{} `json:"hypervisor_version"`
+	}
+
+	err := json.Unmarshal(b, &hypervisor)
+	if err != nil {
+		return err
+	}
+
+	*h = Hypervisor(hypervisor.tmp)
+
+	// Newer versions pass the CPU into around as the correct types, this just needs
+	// converting and copying into place. Older versions pass CPU info around as a string
+	// and can simply be unmarshalled by the json parser
+	switch t := hypervisor.CpuInfo.(type) {
+	case map[string]interface{}:
+		var ok bool
+		if h.CpuInfo.Vendor, ok = t["vendor"].(string); !ok {
+			return fmt.Errorf("CPU info vendor expected to be a string")
+		}
+		if h.CpuInfo.Arch, ok = t["arch"].(string); !ok {
+			return fmt.Errorf("CPU info arch expected to be a string")
+		}
+		if h.CpuInfo.Model, ok = t["model"].(string); !ok {
+			return fmt.Errorf("CPU info model expected to be a string")
+		}
+		if features, ok := t["features"].([]interface{}); ok {
+			for _, feature := range features {
+				if temp, ok := feature.(string); ok {
+					h.CpuInfo.Features = append(h.CpuInfo.Features, temp)
+				} else {
+					return fmt.Errorf("CPU info feaure expected to be a string")
+				}
+			}
+		} else {
+			return fmt.Errorf("CPU info features to be an array")
+		}
+		if topology, ok := t["topology"].(map[string]interface{}); ok {
+			if temp, ok := topology["sockets"].(float64); ok {
+				h.CpuInfo.Topology.Sockets = int(temp)
+			} else {
+				return fmt.Errorf("CPU info topology sockets expected to be numeric")
+			}
+			if temp, ok := topology["cores"].(float64); ok {
+				h.CpuInfo.Topology.Cores = int(temp)
+			} else {
+				return fmt.Errorf("CPU info topology cores expected to be numeric")
+			}
+			if temp, ok := topology["threads"].(float64); ok {
+				h.CpuInfo.Topology.Threads = int(temp)
+			} else {
+				return fmt.Errorf("CPU info topology threads expected to be numeric")
+			}
+		} else {
+			return fmt.Errorf("CPU info topology expected to be a map")
+		}
+	case string:
+		err := json.Unmarshal([]byte(t), &h.CpuInfo)
+		if err != nil {
+			return err
+		}
+	}
+
+	// A feature in OpenStack may return this value as floating point
+	switch t := hypervisor.HypervisorVersion.(type) {
+	case int:
+		h.HypervisorVersion = t
+	case float64:
+		h.HypervisorVersion = int(t)
+	default:
+		return fmt.Errorf("Hypervisor version of unexpected type")
+	}
+
+	return nil
+}
+
+type HypervisorPage struct {
+	pagination.LinkedPageBase
```

Comment:
If this is a linked page, it should have a way to get from one page to the next (usually a `Links` field). Does that exist?
I noticed there are `Marker` and `Limit` fields in the `ListOpts`, so maybe it's a `MarkerPage`?